R/bootstrap-model.R

```diff
@@ -0,0 +1,668 @@
+  fs::file_copy(
+    orig_mod_path, mod_path_ext, overwrite = boot_args$overwrite
+  )
+
+
+  # Sample data and assign new IDs
+  data_new <- resample_df(
+    boot_args$orig_data,
+    key_cols = "ID", # TODO: should this be a user arg?
+    strat_cols = boot_args$strat_cols,
+    replace = TRUE
+  ) %>% dplyr::rename("OID" = "ID", "ID" = "KEY") %>% # TODO: should this be a user arg?
+    dplyr::select(all_of(unique(c(names(boot_args$orig_data), "OID")))) # TODO: should this be a user arg?
+
+  mod_name <- basename(mod_path)
+  orig_mod_id <- boot_args$orig_mod_id
+
+  # Write out new dataset
+  data_boot_name <- glue("boot_{orig_mod_id}_{mod_name}.csv")
+  data_path_boot <- file.path(boot_args$boot_data_dir, data_boot_name)
+  data.table::fwrite(data_new, data_path_boot , na = '.', quote = FALSE)
+
+  # Set Problem and relative datapath (to be sourced in control stream)
+  prob <- glue("Bootstrap run {mod_name} of model {orig_mod_id}")
+  data_path_rel <- fs::path_rel(data_path_boot, boot_args$boot_dir) %>%
+    adjust_data_path_ext(mod_path = mod_path_ext, reverse = TRUE)
+
+  # Create new model object
+  mod <- new_model(
+    mod_path,
+    .overwrite = boot_args$overwrite,
+    .tags = "BOOTSTRAP_RUN"
+  )
+
+  # Overwrite $PROB and $DATA records
+  modify_prob_statement(mod, prob)
+  modify_data_path_ctl(mod, data_path_rel)
+
+  return(mod)
+}
+
+
+#' Store bootstrap run details before submission
+#'
+#' @param boot_runs list of boostrap model objects created by `make_boot_run()`.
+#' @inheritParams make_boot_run
+#'
+#' @details
+#' This is mainly meant to ensure traceability and enhance the portability of
+#' bootstrap "model" objects.
+#'
+#'
+#' @keywords internal
+make_boot_spec <- function(boot_models, boot_args){
+  boot_dir <- boot_args$boot_dir
+  json_path <- file.path(boot_dir, "bbr_boot_spec.json")
+
+  if(fs::file_exists(json_path) && isFALSE(boot_args$overwrite)){
+    rlang::abort(
+      c(
+        glue("A bootstrap specification file already exists at `{json_path}`"),
+        "i" = "Pass `.overwrite = TRUE` to overwrite."
+      )
+    )
+  }
+
+  overall_boot_spec <- list(
+    problem = glue("Bootstrap of {basename(boot_args$orig_mod_path)}"),
+    strat_cols = boot_args$strat_cols,
+    seed = boot_args$seed,
+    n_samples = boot_args$n_samples,
+    model_path = get_model_path(boot_args$boot_run),
+    based_on_model_path = boot_args$orig_mod_path,
+    based_on_data_path = get_data_path(boot_args$boot_run),
+    model_md5 = tools::md5sum(get_model_path(boot_args$boot_run)),
+    based_on_model_md5 = tools::md5sum(boot_args$orig_mod_path),
+    based_on_data_md5 = tools::md5sum(get_data_path(boot_args$boot_run)),
+    output_dir = boot_args$boot_run[[ABS_MOD_PATH]]
+  )
+
+  boot_run_ids <- purrr::map_chr(boot_models, function(boot_run){
+    paste0("run_", basename(boot_run[[ABS_MOD_PATH]]))
+  })
+
+  boot_run_spec <- purrr::map(boot_models, function(boot_run){
```

Comment:
This is a bottle neck; for the analysis data set, it takes 1 min to resample and write the control streams and 35 sec to generate this boot spec. I wonder if there's a simpler way to do this; I think the extra time is taken by the overhead around these functions.

Reply:
I think we can definitely optimize this at some point. Thanks for pointing that out.

Reply:
FWIW I decided to address this ahead of the next dev tag. Was initially worried about the refactor messing with too many functions to risk impacting the timeline, but I saved 95% of the time by just dropping the individual problem statements and data paths from the spec file.